File: app/src/main/java/com/xero/interview/bankrecmatchmaker/MatchAdapter.java
package com.xero.interview.bankrecmatchmaker;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.xero.interview.bankrecmatchmaker.core.common.CurrencyFormatter;
import com.xero.interview.bankrecmatchmaker.databinding.ListItemMatchBinding;

import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

public class MatchAdapter extends ListAdapter<MatchItem, MatchAdapter.ViewHolder> {

    private final OnItemCheckedListener onItemCheckedListener;
    private final CurrencyFormatter currencyFormatter;
    private final Set<MatchItem> selectedItems = new HashSet<>();

    public interface OnItemCheckedListener {
        void onItemChecked(MatchItem item, boolean isChecked);
    }

    public MatchAdapter(OnItemCheckedListener listener) {
        super(DIFF_CALLBACK);
        this.onItemCheckedListener = listener;
        this.currencyFormatter = new CurrencyFormatter(Locale.getDefault(), java.util.Currency.getInstance(Locale.getDefault()));
    }

    public void setItemSelected(MatchItem item, boolean isSelected) {
        if (isSelected) {
            selectedItems.add(item);
        } else {
            selectedItems.remove(item);
        }
        notifyItemChanged(getCurrentList().indexOf(item));
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ListItemMatchBinding binding = ListItemMatchBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        MatchItem item = getItem(position);
        holder.bind(item, selectedItems.contains(item));
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private final ListItemMatchBinding binding;

        ViewHolder(@NonNull ListItemMatchBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(final MatchItem matchItem, boolean isSelected) {
            binding.textMain.setText(matchItem.paidTo());
            binding.textTotal.setText(currencyFormatter.format(matchItem.total()));
            binding.textSubLeft.setText(matchItem.transactionDate());
            binding.textSubRight.setText(matchItem.docType());

            binding.getRoot().setChecked(isSelected);
            binding.getRoot().setOnClickListener(v -> {
                boolean newCheckedState = !binding.getRoot().isChecked();
                setItemSelected(matchItem, newCheckedState);
                onItemCheckedListener.onItemChecked(matchItem, newCheckedState);
            });
        }
    }

    private static final DiffUtil.ItemCallback<MatchItem> DIFF_CALLBACK = new DiffUtil.ItemCallback<MatchItem>() {
        @Override
        public boolean areItemsTheSame(@NonNull MatchItem oldItem, @NonNull MatchItem newItem) {
            return oldItem.equals(newItem);
        }

        @Override
        public boolean areContentsTheSame(@NonNull MatchItem oldItem, @NonNull MatchItem newItem) {
            return oldItem.equals(newItem);
        }
    };

}

File: app/src/main/java/com/xero/interview/bankrecmatchmaker/MatchAdapter.kt
package com.xero.interview.bankrecmatchmaker

import android.view.LayoutInflater
import android.view.ViewGroup

import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView

import com.xero.interview.bankrecmatchmaker.core.common.CurrencyFormatter
import com.xero.interview.bankrecmatchmaker.data.accounting_records.model.AccountingRecord
import com.xero.interview.bankrecmatchmaker.databinding.ListItemMatchBinding
import java.util.Currency

import java.util.Locale

class MatchAdapter(
    private val onItemCheckedListener: (AccountingRecord, Boolean) -> Unit,
    private val canSelectItem: (AccountingRecord) -> Boolean
) : ListAdapter<AccountingRecord, MatchAdapter.ViewHolder>(DIFF_CALLBACK) {

    private val currencyFormatter =
        CurrencyFormatter(Locale.getDefault(), Currency.getInstance(Locale.getDefault()))
    private val selectedItems = mutableSetOf<AccountingRecord>()

    fun setItemSelected(item: AccountingRecord, isSelected: Boolean) {
        if (isSelected && !canSelectItem(item)) {
            return
        }
        if (isSelected) {
            selectedItems.add(item)
        } else {
            selectedItems.remove(item)
        }
        notifyItemChanged(currentList.indexOf(item))
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding =
            ListItemMatchBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = getItem(position)
        val isSelected = selectedItems.contains(item)
        holder.bind(item, isSelected)
    }

    inner class ViewHolder(private val binding: ListItemMatchBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(accountingRecord: AccountingRecord, isSelected: Boolean) {
            binding.textMain.text = accountingRecord.paidTo
            binding.textTotal.text = currencyFormatter.format(accountingRecord.total)
            binding.textSubLeft.text = accountingRecord.transactionDate
            binding.textSubRight.text = accountingRecord.docType

            binding.root.isChecked = isSelected
            binding.root.setOnClickListener {
                val newCheckedState = !binding.root.isChecked
                setItemSelected(accountingRecord, newCheckedState)
                onItemCheckedListener(accountingRecord, newCheckedState)
            }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<AccountingRecord>() {
            override fun areItemsTheSame(
                oldItem: AccountingRecord,
                newItem: AccountingRecord
            ): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(
                oldItem: AccountingRecord,
                newItem: AccountingRecord
            ): Boolean {
                return oldItem == newItem
            }
        }
    }
}